### ISSUE TITLE
Editor oddness

### ISSUE THREAD

**Issue by fake-name** [workflowproducts/postage#355]
Note: I really appreciate what you folks have done here, and I really am not trying to kind of "dump" on you with issues, this is mostly a note of things I noticed and thought might be of interest, rather then any actual problems (and I come across as a ass a lot of the time online despite trying to not do so). Thanks so much for releasing this tool.

------

I've just been playing with postage, and while it's nice, the editor has a lot of *odd* quirks.

When I'm optimizing a query, I do a lot of `EXPLAIN ` or `EXPLAIN ANALYZE` runs. In the SQL editor, this is two clicks (open drop down, click button). I guess this was done to minimize the toolbar size?

If so, why is there an entire button that just tells me about the text control widget being used? If I'm doing SQL stuff, I don't care about the text control. If I'm looking for a javascript text editor widget, I'm probably going to just use google. Why is that even an thing? It'd be much more useful to have separate  `EXPLAIN ` and `EXPLAIN ANALYZE` buttons, and just get rid of the ace button.

<kbd>Ctrl</kbd> + <kbd>W</kbd> seems to close the entire application, rather then the active tab, like is normal (whoops, I've done this a few times).

On that topic, the commands in the pop-up dialog when you click the "Ace" seem to be hit-and-miss, I assume that it's probably related to the current focus. If I click anywhere in the EXPLAINed query section, the editor shortcuts stop working. I understand the mechanism behind this, but it definitely feels weird.

On the topic of stuff that feels weird, when you create a new tab, it's created at the very *left* of the tab bar, rather then on the right-hand side (like every application I've ever used that has tabs does). Again, it's not a problem, it's just off-putting.

Looking around more, holy crap, some of the stuff under the "tools" menu is *awesome*. Maybe see about putting that somewhere more apparent. There's a "History" and "Notify" button, each of which doesn't seem to do much. I've executed a bunch of queries in my current session, and the history is still empty. Maybe add similar buttons for the tools stuff?

Apparently there are keyboard shortcuts for some of the stuff in the editor menu. (<kbd>F7</kbd>, etc...), That's great, why isn't it noted in the relevant menu? The common standard for menus is to have the keyboard shortcut generally in slightly smaller text on the right hand side of the menu item. That helps discoverability enormously, since it's already a known standard for that kind of thing.
 

![2017-08-07 22_42_04-gparted_7_big png 874x612](https://user-images.githubusercontent.com/1401239/29057839-eed2cd54-7bc1-11e7-9c3b-5779c065e256.png)

The query graph is also quite cool, but it seems to have default-zoom issues:

![image](https://user-images.githubusercontent.com/1401239/29057907-3de4291a-7bc2-11e7-89d3-8cc01bfb0b13.png)

It'd probably be ideal to have the graph view default to zoom-to-fit, at least when it's opened for the first time.  
Additionally, the graph seems to be built vertically. This is also kind of sub-optimal, since modern screens are far wider then they are tall. It'd probably let you view it much more easily if it were structured horizontally. 

Also, there is the weird "page curl" thing in the bottom right. What, exactly, is that supposed to be? I clicked on it experimentally, and apparently it's a *different* menu. Why do certain menu options need to be hidden behind a completely new and custom interface cue? If you need an additional menu, put it at the top of the window, where the rest of the menus are, and where you're not inventing new UI symbology.  There's plenty of applications out there for being innovative in UI design, but a tool I'm trusting to manipulate my database is not one of them. A large part of building that trust is being *predictable*, and when I have to experimentally click on things just to figure out what they do, that kind of ruins one's confidence. 

I don't have any actual problems, but the above might be useful feedback.

**Comment by justintocci** [comment]
The "ACE Tips" button lists the keyboard shortcuts. Among those you'll find these:

Use "F7" to run the PostgreSQL Explain function.
Use "Shift-F7" to run the PostgreSQL Explain Analyze function.

But I hear you about the Explain buttons. If you go into the green menu bar and click "Options" and then choose "Postage Options" you can click the "SQL Toolbar Button Style:" to "unlabeled" and you'll get Explain and Explain Analyze buttons as you requested.

I apologize for the inconvenience, but for many months we lost many users because they didn't find the Explain buttons because they were just labeled "E". Occasionally, people would even post issues or ask us in person for Explain. It was crazy.

I hope that solves your issues with Explain. If not, please elaborate what we could do.

**Comment by justintocci** [comment]
> **If I anywhere in the EXPLAINed query section,** the editor shortcuts stop working. I understand the mechanism behind this, but it definitely feels weird.

I'm having trouble parsing the phrase in bold. When exactly do the editor shortcuts stop working? I tried a few things like clicking buttons but I couldn't break it.

**Comment by justintocci** [comment]
I opened issue #356 for "Ctrl + W seems to close the entire application, rather then the active tab"

**Comment by fake-name** [comment]
> I'm having trouble parsing the phrase in bold. When exactly do the editor shortcuts stop working? I tried a few things like clicking buttons but I couldn't break it.

Whoops, dropped a "click". Basically, if you move the focus away from the text box, the shortcuts stop working.

Internally, I'd assume that this is because the shortcuts are routed through the editor window, rather then through the application overall. I understand the cause, but it's kind of quirky.

**Comment by justintocci** [comment]
I opened issue #357 for the tab issue. We just recently did some work on tabs but then did a release kind of in the middle of it so I apologize for that.

**Comment by justintocci** [comment]
?!?!?! 

> I've executed a bunch of queries in my current session, and the history is still empty. 

How is this happening? My history is working, I just did a SELECT and it appeared in history right away. Do you have any javascript errors? What version of Postage are you using? I'm on 3.2.17

**Comment by fake-name** [comment]
> I apologize for the inconvenience, but for many months we lost many users because they didn't find the Explain buttons because they were just labeled "E". Occasionally, people would even post issues or ask us in person for Explain. It was crazy.

Ah, I see. In that case, maybe label the button "help", rather then "Ace?" I knew that "Ace" was a javascript code editor, but that's because I used it in a recent project of my own.

> How is this happening? My history is working, I just did a SELECT and it appeared in history right away. Do you have any javascript errors? What version of Postage are you using? I'm on 3.2.17

Wait, do you consider `EXPLAIN` or `EXPLAIN ANALYZE` queries? Oh, that's the problem. I assumed it'd be anything that had to touch the database, rather then just a subset of it.

**Comment by justintocci** [comment]
I opened #358 to add the Explain shortcuts to the Explain dropdown. We'll also add the shortcuts to the tooltip popups.

**Comment by fake-name** [comment]
cool beans!

**Comment by justintocci** [comment]
I opened #359 to relabel the ACE button. This is so obvious. Thank you so much for pointing it out. We have no excuse for not seeing this sooner.

**Comment by justintocci** [comment]
I opened #360 to add Explain and anything else we're missing to the history. Thank you for all of these items. You are awesome!

**Comment by justintocci** [comment]
> I really appreciate what you folks have done here, and I really am not trying to kind of "dump" on you with issues, this is mostly a note of things I noticed and thought might be of interest, rather then any actual problems (and I come across as a ass a lot of the time online despite trying to not do so). Thanks so much for releasing this tool.

You just keep on dumping issues here my friend. You are very welcome any time.

**Comment by justintocci** [comment]
I opened #361 to address the focus issue on the shortcuts. Addressing this will make it feel more like an application and less like a web app so I'm in favor of it, but I can't promise how quickly we can do it. If it requires a lot of work due to architectural issues it may be delayed or even closed due to lack of resources. Thank you for pointing it out.

**Comment by fake-name** [comment]
A lot of this was less "Fix \<xxx\> now", and more "have you thought about \<thing\>". No hurry.

Once I figured out how the UI worked, I managed to do what I wanted. 

Incidentally, if you're adding tool-tips, it'd be awesome to have some tool-tips for the various columns of the EXPLAIN result. 

Frankly, I read the results, then read the postgres documentation on the explain results, then read some miscellaneous articles, and I think I'm *still* confused, but that's not exactly something an application can fix.
